crates/cubecl-linalg/src/matmul/components/global/load/strategy/async_full_cyclic.rs

```diff
@@ -64,51 +86,94 @@ impl<T: TilingOrder> AsyncFullLoadingStrategy for AsyncFullCyclicLoading<T> {
         };
 
         let num_slices = comptime!(num_slices_per_tile * stage_dim.tile_count());
-        let num_slices_per_unit = num_slices.div_ceil(total_units);
+        let num_tasks = num_slices.div_ceil(total_units);
 
         let unit_id = UNIT_POS_Y * config.plane_dim() + UNIT_POS_X;
 
-        #[unroll(num_slices_per_unit==1)]
-        for i in 0..num_slices_per_unit {
-            let slice_index = unit_id + total_units * i;
-
-            let nth_tile = slice_index / num_slices_per_tile;
-            let (tile_x, tile_y) = ContiguousTilingLayout::<T>::to_x_y::<G::SmmConfig>(
-                nth_tile,
-                input_ident.as_ident(),
-                config.to_smm_config(),
-            );
-            let nth_slice = slice_index % num_slices_per_tile;
-
-            // TODO make branching comptime conditional
-            if slice_index < num_slices {
-                let window = read_view.load_window_in_tile::<G>(
-                    (tile_x, tile_y),
-                    nth_slice,
-                    input_ident,
-                    config,
-                );
-
-                // Where this unit writes source in the stage
-                let slice_destination_offset =
-                    (nth_tile * num_slices_per_tile + nth_slice) * slice_length_in_lines;
-
-                // Make destination start at offset
-                let mut destination = stage.as_slice_mut().slice_mut(
-                    slice_destination_offset,
-                    slice_destination_offset + slice_length_in_lines,
-                );
-
-                CM::memcpy_async(
-                    mechanism,
-                    &window.slice.try_cast_unchecked(),
-                    &mut destination,
-                );
-            }
+        AsyncFullCyclicJob::<MP, CM, T> {
+            unit_id,
+            read_view,
+            stage,
+            mechanism,
+            num_tasks,
+            total_units,
+            num_slices,
+            input_ident,
+            num_slices_per_tile,
+            slice_length_in_lines,
         }
     }
 
     fn barrier_level() -> BarrierLevel {
         BarrierLevel::cube_manual(0u32)
     }
 }
+
+#[derive(CubeType, Clone, Copy)]
+pub struct AsyncFullCyclicJob<MP: MatmulPrecision, CM: CopyMechanism<MP::ES>, T: TilingOrder> {
+    unit_id: u32,
+
+    read_view: TensorReader<MP::EI>,
+    stage: Stage<MP::ES, ContiguousTilingLayout<T>>,
+    mechanism: CM,
+
+    #[cube(comptime)]
+    num_tasks: u32,
+    #[cube(comptime)]
+    total_units: u32,
+    #[cube(comptime)]
+    num_slices: u32,
+    #[cube(comptime)]
+    input_ident: InputIdent,
+    #[cube(comptime)]
+    num_slices_per_tile: u32,
+    #[cube(comptime)]
+    slice_length_in_lines: u32,
+}
+
+#[cube]
+impl<MP: MatmulPrecision, CM: CopyMechanism<MP::ES>, T: TilingOrder> LoadingJob<MP>
+    for AsyncFullCyclicJob<MP, CM, T>
+{
+    fn len(this: &Self) -> u32 {
+        this.num_tasks.runtime()
+    }
```

Comment:
you can use a comptime trait (without cube annotation) and ask for AsyncFullCyclicJob to also implement its comptime version. This way, you will be able to return a u32 directly without converting it to runtime.

An alternative is to move all comptime config in a single struct and implement the comptime function there. With a possible trait for the config. 

Similar to StageConfig or GlobalConfig.